packages/@lwc/engine-core/src/framework/rendering.ts

```diff
@@ -627,16 +645,72 @@ export function allocateChildren(vnode: VCustomElement, vm: VM) {
             );
         }
     }
+
+    // If any of the children being allocated are VFragments, we remove the text delimiters and flatten all immediate
+    // children VFragments to avoid them interfering with default slot behavior.
+
+    // If any of the VFragments were not stable, the children are marked dynamic.
+    let allocatedChildren = children;
+    if (children.some((vnode) => vnode && isVFragment(vnode))) {
+        const { stable, flattenedChildren } = flattenFragmentChildren(children);
+        allocatedChildren = flattenedChildren;
+        vnode.children = flattenedChildren;
+
+        if (!stable) {
+            markAsDynamicChildren(flattenedChildren);
+        }
+    }
+
+    vm.aChildren = allocatedChildren;
+
     if (shadowMode === ShadowMode.Synthetic || renderMode === RenderMode.Light) {
         // slow path
-        allocateInSlot(vm, children, vnode.owner);
+        allocateInSlot(vm, allocatedChildren, vnode.owner);
         // save the allocated children in case this vnode is reused.
-        vnode.aChildren = children;
+        vnode.aChildren = allocatedChildren;
         // every child vnode is now allocated, and the host should receive none directly, it receives them via the shadow!
         vnode.children = EmptyArray;
     }
 }
 
+/**
+ * Flattens the contents of all VFragments in an array of VNodes and removes the text delimiters on those VFragments.
+ * Returns an array of the flattened nodes without text nodes, along with a boolean indicating whether the set of children
+ * can be considered stable.
+ *
+ * Note that with the VFragment delimiters removed, the contents must be treated as dynamic for them to rerender correctly.
+ *
+ * This function is used for slotted VFragments in native shadow mode to avoid the text delimiters interfering with
+ * slotting functionality.
+ */
+function flattenFragmentChildren(children: readonly (VNode | null)[]): {
+    stable: 0 | 1 | undefined;
+    flattenedChildren: (VNode | null)[];
+} {
+    let stable: 0 | 1 | undefined;
```

Comment:
The boolean logic below gets a little confusing to me since there are three possible values. I'd suggest making this an explicit `boolean` rather than `0 | 1 | undefined`.